File: src/main/java/com/coding/programmers/level3/단어변환.java
package com.coding.programmers.level3;

import java.util.*;

public class 단어변환 {
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.solution("hit", "cog", new String[]{
                "hot", "dot", "dog", "lot", "log", "cog"
        }));
    }
    private static class Solution {
        public int solution(String begin, String target, String[] words) {
            Queue<Node> queue = new ArrayDeque<>();
            Set<String> distSet = new HashSet<>();
            queue.add(new Node(begin, 0));
            distSet.add(begin);

            while (!queue.isEmpty()) {
                Node now = queue.poll();

                if (target.equals(now.word)) {
                    return now.weight;
                }

                for (String word : words) {
                    if (!distSet.contains(word) && now.compareWith(word)) {
                        queue.offer(new Node(word, now.weight + 1));
                        distSet.add(word);
                    }
                }
            }
            return 0;
        }

        private static class Node {
            String word;
            int weight;

            public Node(String word, int weight) {
                this.word = word;
                this.weight = weight;
            }

            public boolean compareWith(String target) {
                int count = 0;

                for (int i = 0; i < this.word.length(); i++) {
                    if (this.word.charAt(i) != target.charAt(i) && ++count > 1) {
                        break;
                    }

                }
                return count == 1;
            }
        }
    }
}

File: src/main/kotlin/com/coding/programmers/graph/bfs/단어변환.kt
package com.coding.programmers.graph.bfs

import java.util.ArrayDeque

fun main() {
    val r= 단어변환().solution("hit", "cog", arrayOf("hot", "dot", "dog", "lot", "log", "cog"))
    println(r)
}

/**
 * 한 번에 한 개의 알파벳만 바꿀 수 있다.
 * words에 있는 단어로만 변환 가능
 */
class 단어변환 {
    fun solution(begin: String, target: String, words: Array<String>): Int {
        var answer = 0
        val queue = ArrayDeque<Node>()
        val dist = HashSet<String>()

        queue.add(Node(begin, 0))
        dist.add(begin)

        while(queue.isNotEmpty()) {
            val dq = queue.poll();

            if (dq.word == target) {
                answer = dq.weight
                break;
            }

            for (word in words) {
                if (!dist.contains(word) && dq.compareWith(word)) {
                    queue.offer(Node(word, dq.weight + 1))
                    dist.add(word)
                }
            }
        }

        return answer
    }
    data class Node(
        val word: String,
        val weight: Int,
    ) {
        fun compareWith(target: String): Boolean {
            var count = 0

            for (i in word.indices) {
                if (word[i] != target[i] && ++count > 1) {
                    break
                }
            }
            return count == 1
        }
    }
}